### ISSUE TITLE
Bug: [no-unused-vars] false positive on `infer` generic nested within conditional type

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#6253]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=C4TwDgpgBA+gYgewQHgCoD4oF4qqhAD2AgDsATAZygAoBYAKCl3yNMqgFcSBrEhAdxJQA-FACWJAGYQATlABKUAFxQSEAG6yGASgZMmo+XuWqNsoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWwOzKIAm1AG4BDaKgwBtSImjQ20SABpMkEWICSAc3bREABWG5cMhuigA9APqQAvgF0bIG0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
type _Foo<T> = T extends (
  T extends unknown ? infer R : never
)
    ? R
  : never
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": ["error", {"varsIgnorePattern": "^_"}]
  },
};
```

### Expected Result

no error because `R` is used

### Actual Result

```
'R' is defined but never used. Allowed unused vars must match /^_/u. 2:29 - 2:30
```

**Comment by bradzacher** [comment]
The issue is that `infer` is quite difficult to properly replicate in terms of how it declares a variable.

In this particular case `infer R` declares a variable in the outer conditional type, but if it were instead in `extends` clause it would only create a variable local to the inner conditional type.

Theres also things like `infer` doesn't create variables that can be used in the "other" branch of the conditional.

Conditional types are just really weird.

---

Question: why are you writing this gnarly code with a conditional as the `extends`? I'd love to see the real code you were writing to create that.
I've seen nested conditionals before - but they are always in the branches, not the extends clause.

**Comment by DetachHead** [comment]
> why are you writing this gnarly code with a conditional as the `extends`? I'd love to see the real code you were writing to create that.

it's a minified version of this:
```ts
type Wrapped<T> = T extends (
    T extends {} ? NonNullable<infer R> : infer R
)
    ? {
          [P in keyof R]: R[P]
      }
    : never
```
which is a hacky way to remove unnessecary intersections from types without changing their structure (eg. turns `number & {}` into `number` and `{foo: number} & {bar: number}` into `{foo: number, bar: number}`).

this type is then used to compare two types to see whether they're equal:
```ts
type Equals<A, B> = 
    (<T>() => T extends Wrapped<A> ? 1 : 2) extends <T>() => T extends Wrapped<B> ? 1 : 2
        ? true
        : false
```

without this `Wrapped` type, the following types would be incorrectly reported as `false`:
```ts
type Foo = Equals<number & {}, number> // true
type Bar = Equals<{foo: number} & {bar: number}, {foo: number, bar: number}> // true
```
see https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-778623742
[playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA6gTgQzJAJgHgCoD4oF4oZQQAewEAdigM5QAUAsAFBQFGkXVQDeAvlAPxQAcgHtyQgK4AbKQgBGUiGgCW5AGYQ4UAEo4AXFFUat2pgEomLQV0ssoAbQAKh8lADWEECLU6Auge0nX1seWwNyCAA3TSYmUEgoAFEARwkEKSo0AEEAGigAIRxcW1pMLFozPBxCEjJKGngkVGycQQBGKAMAJkrajhoyiqrWPvrYRGQIdEKBKA7u20FgOAkIMKg1dKo1xjjwaAAxERE8JNSttHIJAFs5TSgAMm4ePKvbzSwmAHovuwA9fh7BL5BBafApNIZNBcNTHcI3O5wPhPLhyUHw95IvIwuFQN6IvJouAYxE8T6MH7-fhAA)

**Comment by ArrayKnight** [comment]
Running into a similar issue where we're using infer to gain access to the second parameter in a generic:
https://typescript-eslint.io/play/#ts=5.1.6&showAST=es&fileType=.ts&code=PTAEDEEMGsFNQJYFsAOAbWTYDsAulcEB7bAKFwE8V4A1BWAdwB4AVAGlAFUA%2BUAXlAtQAMi6lyVeAFEAHrgBOkAMa46jAAryiKAM6teAobDk4AJjtBrmCbADNY80AH0ON%2B4-W8A-KHWgAXKDYsABuDgDcpEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXWZRAE0IDcBDaVDSfAMxcgBpxtIABAFwCeABxQBjaAEth-QiniTc-APQFipclToN0URNGj5okbgF8QpoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

```
// Fake implementation
type View<T, U> = T & U

type ExtractViewProps<T> = T extends View<infer _, infer P> ? P : never;
```

We have no control over the `View` type, and the first inferred parameter is of no use to us. We don't want to enable the `varsIgnorePattern` option, but also feel like this should either be seen as an arg (and ignored via our `argsIgnorePattern`, or have a new option that is type specific

**Comment by bradzacher** [comment]
@ArrayKnight thats unrelated to this issue and is not a false positive. You actually have a completely unused variable there. If you don't care about the type then just use `any` - it has exactly the same behaviour as what you've written without generating an unused variable.
